crates/isograph_compiler/src/compiler_state.rs

```diff
@@ -0,0 +1,134 @@
+    ///   (called artifacts). This step should not fail. It should panic if any
+    ///   invariant is violated, or represent that invariant in the type system.
+    /// - Delete and recreate the artifacts on disk.
+    ///
+    /// ## Additional things we do
+    ///
+    /// In addition to the things we do above, we also do some specific things like:
+    ///
+    /// - if a client field is defined on an interface, add it to each concrete
+    ///   type. So, if User implements Actor, you can define Actor.NameDisplay, and
+    ///   select User.NameDisplay
+    /// - create fields from exposeAs directives
+    ///
+    /// These are less "core" to the overall mission, and thus invite the question
+    /// of whether they belong in this function, or at all.
+    ///
+    /// ## Sequentially written vs Salsa architecture
+    ///
+    /// Isograph is currently written in a fairly sequential fashion, e.g.:
+    ///
+    /// let result_1 = step_1()?;
+    /// let result_2 = step_2()?;
+    /// step_3(result_1, result_2)?;
+    ///
+    /// Where each step is completed before the next one starts. This has advantages:
+    /// namely, it is easy to read. But, we most likely want to report all the errors
+    /// we can (i.e. from both step_1 and step_2), rather than just the first error
+    /// encountered (i.e. just step_1).
+    ///
+    /// In the long term, we want to describe everything as a tree, e.g.
+    /// `step_3 -> [step_1, step_2]`, and this will "naturally" parallelize everything.
+    /// This is also necessary to adopt a Rust Analyzer-like (Salsa) architecture, which is
+    /// important for language server performance. In a Salsa architecture, we invalidate
+    /// leaves (e.g. a given file changed), and invalidate everything that depends on that
+    /// leaf. Then, when we need a result (e.g. the errors to show on a given file), we
+    /// re-evaluate (or re-use the cached value) of everything from that result on down.
+    pub fn batch_compile(self) -> Result<CompilationStats, BatchCompileError> {
+        let source_files = SourceFiles::new(&self.config)?;
+        let stats = source_files.contains_iso.stats();
+        let (_, artifacts) = source_files.validate_and_create_artifacts(&self.config)?;
+        let total_artifacts_written =
+            write_artifacts_to_disk(artifacts, &self.config.artifact_directory)?;
+        Ok(CompilationStats {
+            client_field_count: stats.client_field_count,
+            entrypoint_count: stats.entrypoint_count,
+            total_artifacts_written,
+        })
+    }
+
+    pub fn compile(&mut self) -> Result<CompilationStats, BatchCompileError> {
+        let source_files = SourceFiles::new(&self.config)?;
+        let stats = source_files.contains_iso.stats();
+        self.source_files = Some(source_files.clone());
+        let (_, artifacts) = source_files.validate_and_create_artifacts(&self.config)?;
```

Comment:
should we extract a common fn:
```
fn validate_and_create_artifacts_from_source_files(
  source_files: SourceFiles,
  artifact_directory,
): usize /* artifact count */
```

Reply:
Yeah, we can do it for now and change later if needed to keep schema and artifact hashes in the state.